File: pyrasterframes/python/pyrasterframes/rasterfunctions.py
"""
This module creates explicit Python functions that map back to the existing Scala
implementations. Most functions are standard Column functions, but those with unique
signatures are handled here as well.
"""


from __future__ import absolute_import
from pyspark.sql.types import *
from pyspark.sql.column import Column, _to_java_column
from .context import RFContext


THIS_MODULE = 'pyrasterframes'


def _context_call(name, *args):
    f = RFContext.active().lookup(name)
    return f(*args)


def _celltype(cellTypeStr):
    """ Convert the string cell type to the expected CellType object."""
    return _context_call('cellType', cellTypeStr)


def _create_assembleTile():
    """ Create a function mapping to the Scala implementation."""
    def _(colIndex, rowIndex, cellData, numCols, numRows, cellType):
        jfcn = RFContext.active().lookup('assembleTile')
        return Column(jfcn(_to_java_column(colIndex), _to_java_column(rowIndex), _to_java_column(cellData), numCols, numRows, _celltype(cellType)))
    _.__name__ = 'assembleTile'
    _.__doc__ = "Create a Tile from  a column of cell data with location indices"
    _.__module__ = THIS_MODULE
    return _


def _create_arrayToTile():
    """ Create a function mapping to the Scala implementation."""
    def _(arrayCol, numCols, numRows):
        jfcn = RFContext.active().lookup('arrayToTile')
        return Column(jfcn(_to_java_column(arrayCol), numCols, numRows))
    _.__name__ = 'arrayToTile'
    _.__doc__ = "Convert array in `arrayCol` into a Tile of dimensions `numCols` and `numRows'"
    _.__module__ = THIS_MODULE
    return _


def _create_convertCellType():
    """ Create a function mapping to the Scala implementation."""
    def _(tileCol, cellType):
        jfcn = RFContext.active().lookup('convertCellType')
        return Column(jfcn(_to_java_column(tileCol), _celltype(cellType)))
    _.__name__ = 'convertCellType'
    _.__doc__ = "Convert the numeric type of the Tiles in `tileCol`"
    _.__module__ = THIS_MODULE
    return _


def _create_makeConstantTile():
    """ Create a function mapping to the Scala implementation."""
    def _(value, cols, rows, cellType):
        jfcn = RFContext.active().lookup('makeConstantTile')
        return Column(jfcn(value, cols, rows, cellType))
    _.__name__ = 'makeConstantTile'
    _.__doc__ = "Constructor for constant tile column"
    _.__module__ = THIS_MODULE
    return _


def _create_tileZeros():
    """ Create a function mapping to the Scala implementation."""
    def _(cols, rows, cellType = 'float64'):
        jfcn = RFContext.active().lookup('tileZeros')
        return Column(jfcn(cols, rows, cellType))
    _.__name__ = 'tileZeros'
    _.__doc__ = "Create column of constant tiles of zero"
    _.__module__ = THIS_MODULE
    return _


def _create_tileOnes():
    """ Create a function mapping to the Scala implementation."""
    def _(cols, rows, cellType = 'float64'):
        jfcn = RFContext.active().lookup('tileOnes')
        return Column(jfcn(cols, rows, cellType))
    _.__name__ = 'tileOnes'
    _.__doc__ = "Create column of constant tiles of one"
    _.__module__ = THIS_MODULE
    return _


def _create_rasterize():
    """ Create a function mapping to the Scala rasterize function. """
    def _(geometryCol, boundsCol, valueCol, numCols, numRows):
        jfcn = RFContext.active().lookup('rasterize')
        return Column(jfcn(_to_java_column(geometryCol), _to_java_column(boundsCol), _to_java_column(valueCol), numCols, numRows))
    _.__name__ = 'rasterize'
    _.__doc__ = 'Create a tile where cells in the grid defined by cols, rows, and bounds are filled with the given value.'
    _.__module__ = THIS_MODULE
    return _


def _create_reproject_geometry():
    """ Create a function mapping to the Scala reprojectGeometry function. """
    def _(geometryCol, srcCRSName, dstCRSName):
        jfcn = RFContext.active().lookup('reprojectGeometry')
        return Column(jfcn(_to_java_column(geometryCol), srcCRSName, dstCRSName))
    _.__name__ = 'reprojectGeometry'
    _.__doc__ = """Reproject a column of geometry given the CRS names of the source and destination.
Currently supported registries are EPSG, ESRI, WORLD, NAD83, & NAD27.
An example of a valid CRS name is EPSG:3005.
"""
    _.__module__ = THIS_MODULE
    return _


def _create_explode_tiles():
    """ Create a function mapping to Scala explodeTiles function """
    def _(*args):
        jfcn = RFContext.active().lookup('explodeTiles')
        jcols = [_to_java_column(arg) for arg in args]
        return Column(jfcn(RFContext.active().list_to_seq(jcols)))
    _.__name__ = 'explodeTiles'
    _.__doc__ = 'Create a row for each cell in Tile.'
    _.__module__ = THIS_MODULE
    return _


def _create_explode_tiles_sample():
    """ Create a function mapping to Scala explodeTilesSample function"""
    def _(sample_frac, seed, *tile_cols):
        jfcn = RFContext.active().lookup('explodeTilesSample')
        jcols = [_to_java_column(arg) for arg in tile_cols]
        return Column(jfcn(sample_frac, seed, RFContext.active().list_to_seq(jcols)))

    _.__name__ = 'explodeTilesSample'
    _.__doc__ = 'Create a row for a sample of cells in Tile columns.'
    _.__module__ = THIS_MODULE
    return _


def _create_maskByValue():
    """ Create a function mapping to Scala maskByValue function """
    def _(data_tile, mask_tile, mask_value):
        jfcn = RFContext.active().lookup('maskByValue')
        return Column(jfcn(_to_java_column(data_tile), _to_java_column(mask_tile), _to_java_column(mask_value)))
    _.__name__ = 'maskByValue'
    _.__doc__ = 'Generate a tile with the values from the data tile, but where cells in the masking tile contain the masking value, replace the data value with NODATA.'
    _.__module__ = THIS_MODULE
    return _


_rf_unique_functions = {
    'arrayToTile': _create_arrayToTile(),
    'assembleTile': _create_assembleTile(),
    'cellTypes': lambda: _context_call('cellTypes'),
    'convertCellType': _create_convertCellType(),
    'explodeTiles': _create_explode_tiles(),
    'explodeTilesSample': _create_explode_tiles_sample(),
    'makeConstantTile': _create_makeConstantTile(),
    'maskByValue': _create_maskByValue(),
    'rasterize': _create_rasterize(),
    'reprojectGeometry': _create_reproject_geometry(),
    'tileOnes': _create_tileOnes(),
    'tileZeros': _create_tileZeros(),
}


_rf_column_scalar_functions = {
    'withNoData': 'Assign a `NoData` value to the Tiles in the given Column.',
    'localAddScalar': 'Add a scalar to a Tile',
    'localAddScalarInt': 'Add a scalar to a Tile',
    'localSubtractScalar': 'Subtract a scalar from a Tile',
    'localSubtractScalarInt': 'Subtract a scalar from a Tile',
    'localMultiplyScalar': 'Multiply a Tile by a scalar',
    'localMultiplyScalarInt': 'Multiply a Tile by a scalar',
    'localDivideScalar': 'Divide a Tile by a scalar',
    'localDivideScalarInt': 'Divide a Tile by a scalar',
    'localLessScalar': 'Return a Tile with values equal 1 if the cell is less than a scalar, otherwise 0',
    'localLessScalarInt': 'Return a Tile with values equal 1 if the cell is less than a scalar, otherwise 0',
    'localLessEqualScalar': 'Return a Tile with values equal 1 if the cell is less than or equal to a scalar, otherwise 0',
    'localLessEqualScalarInt': 'Return a Tile with values equal 1 if the cell is less than or equal to a scalar, otherwise 0',
    'localGreaterScalar': 'Return a Tile with values equal 1 if the cell is greater than a scalar, otherwise 0',
    'localGreaterScalarInt': 'Return a Tile with values equal 1 if the cell is greater than a scalar, otherwise 0',
    'localGreaterEqualScalar': 'Return a Tile with values equal 1 if the cell is greater than or equal to a scalar, otherwise 0',
    'localGreaterEqualScalarInt': 'Return a Tile with values equal 1 if the cell is greater than or equal to a scalar, otherwise 0',
    'localEqualScalar': 'Return a Tile with values equal 1 if the cell is equal to a scalar, otherwise 0',
    'localEqualScalarInt': 'Return a Tile with values equal 1 if the cell is equal to a scalar, otherwise 0',
    'localUnequalScalar': 'Return a Tile with values equal 1 if the cell is not equal to a scalar, otherwise 0',
    'localUnequalScalarInt': 'Return a Tile with values equal 1 if the cell is not equal to a scalar, otherwise 0',
}


_rf_column_functions = {
    # ------- RasterFrames functions -------
    'tileDimensions': 'Query the number of (cols, rows) in a Tile.',
    'envelope': 'Extracts the bounding box (envelope) of the geometry.',
    'tileToIntArray': 'Flattens Tile into an array of integers.',
    'tileToDoubleArray': 'Flattens Tile into an array of doubles.',
    'cellType': 'Extract the Tile\'s cell type',
    'aggHistogram': 'Compute the full column aggregate floating point histogram',
    'aggStats': 'Compute the full column aggregate floating point statistics',
    'aggMean': 'Computes the column aggregate mean',
    'aggDataCells': 'Computes the number of non-NoData cells in a column',
    'aggNoDataCells': 'Computes the number of NoData cells in a column',
    'tileHistogram': 'Compute the Tile-wise histogram',
    'tileMean': 'Compute the Tile-wise mean',
    'tileSum': 'Compute the Tile-wise sum',
    'tileMin': 'Compute the Tile-wise minimum',
    'tileMax': 'Compute the Tile-wise maximum',
    'tileStats': 'Compute the Tile-wise floating point statistics',
    'renderAscii': 'Render ASCII art of tile',
    'noDataCells': 'Count of NODATA cells',
    'dataCells': 'Count of cells with valid data',
    'localAdd': 'Add two Tiles',
    'localSubtract': 'Subtract two Tiles',
    'localMultiply': 'Multiply two Tiles',
    'localDivide': 'Divide two Tiles',
    'normalizedDifference': 'Compute the normalized difference of two tiles',
    'localAggStats': 'Compute cell-local aggregate descriptive statistics for a column of Tiles.',
    'localAggMax': 'Compute the cell-wise/local max operation between Tiles in a column.',
    'localAggMin': 'Compute the cellwise/local min operation between Tiles in a column.',
    'localAggMean': 'Compute the cellwise/local mean operation between Tiles in a column.',
    'localAggDataCells': 'Compute the cellwise/local count of non-NoData cells for all Tiles in a column.',
    'localAggNoDataCells': 'Compute the cellwise/local count of NoData cells for all Tiles in a column.',
    'mask': 'Where the mask (second) tile contains NODATA, replace values in the source (first) tile with NODATA.',
    'inverseMask': 'Where the mask (second) tile DOES NOT contain NODATA, replace values in the source (first) tile with NODATA.',
    'localLess': 'Cellwise less than comparison between two tiles',
    'localLessEqual': 'Cellwise less than or equal to comparison between two tiles',
    'localGreater': 'Cellwise greater than comparison between two tiles',
    'localGreaterEqual': 'Cellwise greater than or equal to comparison between two tiles',
    'localEqual': 'Cellwise equality comparison between two tiles',
    'localUnequal': 'Cellwise inequality comparison between two tiles',
    # ------- JTS functions -------
    # spatial constructors
    'st_geomFromGeoHash': '',
    'st_geomFromWKT': '',
    'st_geomFromWKB': '',
    'st_lineFromText': '',
    'st_makeBox2D': '',
    'st_makeBBox': '',
    'st_makePolygon': '',
    'st_makePoint': '',
    'st_makeLine': '',
    'st_makePointM': '',
    'st_mLineFromText': '',
    'st_mPointFromText': '',
    'st_mPolyFromText': '',
    'st_point': '',
    'st_pointFromGeoHash': '',
    'st_pointFromText': '',
    'st_pointFromWKB': '',
    'st_polygon': '',
    'st_polygonFromText': '',
    # spatial converters
    'st_castToPoint': '',
    'st_castToPolygon': '',
    'st_castToLineString': '',
    'st_byteArray': '',
    # spatial accessors
    'st_boundary': '',
    'st_coordDim': '',
    'st_dimension': '',
    'st_envelope': '',
    'st_exteriorRing': '',
    'st_geometryN': '',
    'st_geometryType': '',
    'st_interiorRingN': '',
    'st_isClosed': '',
    'st_isCollection': '',
    'st_isEmpty': '',
    'st_isRing': '',
    'st_isSimple': '',
    'st_isValid': '',
    'st_numGeometries': '',
    'st_numPoints': '',
    'st_pointN': '',
    'st_x': '',
    'st_y': '',
    # spatial outputs
    'st_asBinary': '',
    'st_asGeoJSON': '',
    'st_asLatLonText': '',
    'st_asText': '',
    'st_geoHash': '',
    # spatial processors
    'st_bufferPoint': '',
    'st_antimeridianSafeGeom': '',
    # spatial relations
    'st_translate': '',
    'st_contains': '',
    'st_covers': '',
    'st_crosses': '',
    'st_disjoint': '',
    'st_equals': '',
    'st_intersects': '',
    'st_overlaps': '',
    'st_touches': '',
    'st_within': '',
    'st_relate': '',
    'st_relateBool': '',
    'st_area': '',
    'st_closestPoint': '',
    'st_centroid': '',
    'st_distance': '',
    'st_distanceSphere': '',
    'st_length': '',
    'st_aggregateDistanceSphere': '',
    'st_lengthSphere': '',
}


__all__ = list(_rf_column_functions.keys()) + \
          list(_rf_column_scalar_functions.keys()) + \
          list(_rf_unique_functions.keys())


def _create_column_function(name, doc=""):
    """ Create a mapping to Scala UDF for a column function by name"""
    def _(*args):
        jfcn = RFContext.active().lookup(name)
        jcols = [_to_java_column(arg) for arg in args]
        return Column(jfcn(*jcols))
    _.__name__ = name
    _.__doc__ = doc
    _.__module__ = THIS_MODULE
    return _


def _create_columnScalarFunction(name, doc=""):
    """ Create a mapping to Scala UDF for a (column, scalar) -> column function by name"""
    def _(col, scalar):
        jfcn = RFContext.active().lookup(name)
        return Column(jfcn(_to_java_column(col), scalar))
    _.__name__ = name
    _.__doc__ = doc
    _.__module__ = THIS_MODULE
    return _


def _register_functions():
    """ Register each function in the scope"""
    for name, doc in _rf_column_functions.items():
        globals()[name] = _create_column_function(name, doc)

    for name, doc in _rf_column_scalar_functions.items():
        globals()[name] = _create_columnScalarFunction(name, doc)

    for name, func in _rf_unique_functions.items():
        globals()[name] = func


_register_functions()

File: core/src/main/scala/astraea/spark/rasterframes/RasterFunctions.scala
/*
 * This software is licensed under the Apache 2 license, quoted below.
 *
 * Copyright 2017 Astraea, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     [http://www.apache.org/licenses/LICENSE-2.0]
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package astraea.spark.rasterframes

import astraea.spark.rasterframes.expressions.TileAssembler
import astraea.spark.rasterframes.expressions.accessors._
import astraea.spark.rasterframes.expressions.aggstats._
import astraea.spark.rasterframes.expressions.generators._
import astraea.spark.rasterframes.expressions.localops._
import astraea.spark.rasterframes.expressions.tilestats._
import astraea.spark.rasterframes.expressions.transformers._
import astraea.spark.rasterframes.stats.{CellHistogram, CellStatistics}
import astraea.spark.rasterframes.{functions => F}
import geotrellis.proj4.CRS
import geotrellis.raster.mapalgebra.local.LocalTileBinaryOp
import geotrellis.raster.{CellType, Tile}
import geotrellis.vector.Extent
import org.apache.spark.annotation.Experimental
import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.locationtech.jts.geom.{Envelope, Geometry}

/**
 * UDFs for working with Tiles in Spark DataFrames.
 *
 * @since 4/3/17
 */
trait RasterFunctions {
  import util._

  // format: off
  /** Create a row for each cell in Tile. */
  def explode_tiles(cols: Column*): Column = explode_tiles_sample(1.0, None, cols: _*)

  /** Create a row for each cell in Tile with random sampling and optional seed. */
  def explode_tiles_sample(sampleFraction: Double, seed: Option[Long], cols: Column*): Column =
    ExplodeTiles(sampleFraction, seed, cols)

  /** Create a row for each cell in Tile with random sampling (no seed). */
  def explode_tiles_sample(sampleFraction: Double, cols: Column*): Column =
    ExplodeTiles(sampleFraction, None, cols)

  /** Query the number of (cols, rows) in a Tile. */
  def tile_dimensions(col: Column): Column = GetDimensions(col)

  /** Extracts the bounding box of a geometry as a JTS envelope. */
  @deprecated("Replace usages of this with `st_extent`", "11/4/2018")
  def envelope(col: Column): TypedColumn[Any, Envelope] = GetEnvelope(col)

  /** Extracts the bounding box of a geometry as an Extent */
  def st_extent(col: Column): TypedColumn[Any, Extent] = GeometryToExtent(col)

  /** Flattens Tile into a double array. */
  def tile_to_array_double(col: Column): TypedColumn[Any, Array[Double]] =
    TileToArrayDouble(col)

  /** Flattens Tile into an integer array. */
  def tile_to_array_int(col: Column): TypedColumn[Any, Array[Double]] =
    TileToArrayDouble(col)

  @Experimental
  /** Convert array in `arrayCol` into a Tile of dimensions `cols` and `rows`*/
  def array_to_tile(arrayCol: Column, cols: Int, rows: Int) = withAlias("array_to_tile", arrayCol)(
    udf[Tile, AnyRef](F.arrayToTile(cols, rows)).apply(arrayCol)
  )

  /** Create a Tile from a column of cell data with location indexes and preform cell conversion. */
  def assemble_tile(columnIndex: Column, rowIndex: Column, cellData: Column, tileCols: Int, tileRows: Int, ct: CellType): TypedColumn[Any, Tile] =
    convert_cell_type(TileAssembler(columnIndex, rowIndex, cellData, lit(tileCols), lit(tileRows)), ct).as(cellData.columnName).as[Tile](singlebandTileEncoder)

  /** Create a Tile from  a column of cell data with location indexes. */
  def assemble_tile(columnIndex: Column, rowIndex: Column, cellData: Column, tileCols: Column, tileRows: Column): TypedColumn[Any, Tile] =
    TileAssembler(columnIndex, rowIndex, cellData, tileCols, tileRows)

  /** Extract the Tile's cell type */
  def cell_type(col: Column): TypedColumn[Any, CellType] = GetCellType(col)

  /** Change the Tile's cell type */
  def convert_cell_type(col: Column, cellType: CellType): TypedColumn[Any, Tile] =
    SetCellType(col, cellType)

  /** Change the Tile's cell type */
  def convert_cell_type(col: Column, cellTypeName: String): TypedColumn[Any, Tile] =
    SetCellType(col, cellTypeName)

  /** Convert a bounding box structure to a Geometry type. Intented to support multiple schemas. */
  def bounds_geometry(bounds: Column): TypedColumn[Any, Geometry] = BoundsToGeometry(bounds)

  /** Assign a `NoData` value to the Tiles. */
  def with_no_data(col: Column, nodata: Double): TypedColumn[Any, Tile] = withAlias("with_no_data", col)(
    udf[Tile, Tile](F.withNoData(nodata)).apply(col)
  ).as[Tile]

  /**  Compute the full column aggregate floating point histogram. */
  def agg_approx_histogram(col: Column): TypedColumn[Any, CellHistogram] =
    HistogramAggregate(col)

  /** Compute the full column aggregate floating point statistics. */
  def agg_stats(col: Column): TypedColumn[Any, CellStatistics] =
    CellStatsAggregate(col)

  /** Computes the column aggregate mean. */
  def agg_mean(col: Column) = CellMeanAggregate(col)

  /** Computes the number of non-NoData cells in a column. */
  def agg_data_cells(col: Column): TypedColumn[Any, Long] = CellCountAggregate.DataCells(col)

  /** Computes the number of NoData cells in a column. */
  def agg_no_data_cells(col: Column): TypedColumn[Any, Long] = CellCountAggregate.NoDataCells(col)

  /** Compute the Tile-wise mean */
  def tile_mean(col: Column): TypedColumn[Any, Double] =
    TileMean(col)

  /** Compute the Tile-wise sum */
  def tile_sum(col: Column): TypedColumn[Any, Double] =
    Sum(col)

  /** Compute the minimum cell value in tile. */
  def tile_min(col: Column): TypedColumn[Any, Double] =
    TileMin(col)

  /** Compute the maximum cell value in tile. */
  def tile_max(col: Column): TypedColumn[Any, Double] =
    TileMax(col)

  /** Compute TileHistogram of Tile values. */
  def tile_histogram(col: Column): TypedColumn[Any, CellHistogram] =
    TileHistogram(col)

  /** Compute statistics of Tile values. */
  def tile_stats(col: Column): TypedColumn[Any, CellStatistics] =
    TileStats(col)

  /** Counts the number of non-NoData cells per Tile. */
  def data_cells(tile: Column): TypedColumn[Any, Long] =
    DataCells(tile)

  /** Counts the number of NoData cells per Tile. */
  def no_data_cells(tile: Column): TypedColumn[Any, Long] =
    NoDataCells(tile)

  def is_no_data_tile(tile: Column): TypedColumn[Any, Boolean] =
    IsNoDataTile(tile)

  /** Compute cell-local aggregate descriptive statistics for a column of Tiles. */
  def agg_local_stats(col: Column) =
    LocalStatsAggregate(col)

  /** Compute the cell-wise/local max operation between Tiles in a column. */
  def agg_local_max(col: Column): TypedColumn[Any, Tile] = LocalTileOpAggregate.LocalMaxUDAF(col)

  /** Compute the cellwise/local min operation between Tiles in a column. */
  def agg_local_min(col: Column): TypedColumn[Any, Tile] = LocalTileOpAggregate.LocalMinUDAF(col)

  /** Compute the cellwise/local mean operation between Tiles in a column. */
  def agg_local_mean(col: Column): TypedColumn[Any, Tile] = LocalMeanAggregate(col)

  /** Compute the cellwise/local count of non-NoData cells for all Tiles in a column. */
  def agg_local_data_cells(col: Column): TypedColumn[Any, Tile] = LocalCountAggregate.LocalDataCellsUDAF(col)

  /** Compute the cellwise/local count of NoData cells for all Tiles in a column. */
  def agg_local_no_data_cells(col: Column): TypedColumn[Any, Tile] = LocalCountAggregate.LocalNoDataCellsUDAF(col)

  /** Cellwise addition between two Tiles or Tile and scalar column. */
  def local_add(left: Column, right: Column): TypedColumn[Any, Tile] = Add(left, right)

  /** Cellwise addition of a scalar value to a tile. */
  def local_add[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] = Add(tileCol, value)

  /** Cellwise subtraction between two Tiles. */
  def local_subtract(left: Column, right: Column): TypedColumn[Any, Tile] = Subtract(left, right)

  /** Cellwise subtraction of a scalar value from a tile. */
  def local_subtract[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile]  = Subtract(tileCol, value)

  /** Cellwise multiplication between two Tiles. */
  def local_multiply(left: Column, right: Column): TypedColumn[Any, Tile] = Multiply(left, right)

  /** Cellwise multiplication of a tile by a scalar value. */
  def local_multiply[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] = Multiply(tileCol, value)

  /** Cellwise division between two Tiles. */
  def local_divide(left: Column, right: Column): TypedColumn[Any, Tile] = Divide(left, right)

  /** Cellwise division of a tile by a scalar value. */
  def local_divide[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] = Divide(tileCol, value)

  /** Perform an arbitrary GeoTrellis `LocalTileBinaryOp` between two Tile columns. */
  def local_algebra(op: LocalTileBinaryOp, left: Column, right: Column):
  TypedColumn[Any, Tile] =
    withAlias(opName(op), left, right)(
      udf[Tile, Tile, Tile](op.apply).apply(left, right)
    ).as[Tile]

  /** Compute the normalized difference of two tile columns */
  def normalized_difference(left: Column, right: Column) =
    NormalizedDifference(left, right)

  /** Constructor for constant tile column */
  def make_constant_tile(value: Number, cols: Int, rows: Int, cellType: String): TypedColumn[Any, Tile] =
    udf(() => F.makeConstantTile(value, cols, rows, cellType)).apply().as(s"constant_$cellType").as[Tile]

  /** Alias for column of constant tiles of zero */
  def tile_zeros(cols: Int, rows: Int, cellType: String = "float64"): TypedColumn[Any, Tile] =
    udf(() => F.tileZeros(cols, rows, cellType)).apply().as(s"zeros_$cellType").as[Tile]

  /** Alias for column of constant tiles of one */
  def tile_ones(cols: Int, rows: Int, cellType: String = "float64"): TypedColumn[Any, Tile] =
    udf(() => F.tileOnes(cols, rows, cellType)).apply().as(s"ones_$cellType").as[Tile]

  /** Where the mask tile contains NODATA, replace values in the source tile with NODATA */
  def mask(sourceTile: Column, maskTile: Column): TypedColumn[Any, Tile] =
    Mask.MaskByDefined(sourceTile, maskTile)

  /** Where the mask tile equals the mask value, replace values in the source tile with NODATA */
  def mask_by_value(sourceTile: Column, maskTile: Column, maskValue: Column): TypedColumn[Any, Tile] =
    Mask.MaskByValue(sourceTile, maskTile, maskValue)

  /** Where the mask tile DOES NOT contain NODATA, replace values in the source tile with NODATA */
  def inverse_mask(sourceTile: Column, maskTile: Column): TypedColumn[Any, Tile] =
    Mask.InverseMaskByDefined(sourceTile, maskTile)

  /** Create a tile where cells in the grid defined by cols, rows, and bounds are filled with the given value. */
  def rasterize(geometry: Column, bounds: Column, value: Column, cols: Int, rows: Int): TypedColumn[Any, Tile] =
    withAlias("rasterize", geometry)(
      udf(F.rasterize(_: Geometry, _: Geometry, _: Int, cols, rows)).apply(geometry, bounds, value)
    ).as[Tile]

  /** Reproject a column of geometry from one CRS to another. */
  def reproject_geometry(sourceGeom: Column, srcCRS: CRS, dstCRSCol: Column): TypedColumn[Any, Geometry] =
    ReprojectGeometry(sourceGeom, srcCRS, dstCRSCol)

  /** Reproject a column of geometry from one CRS to another. */
  def reproject_geometry(sourceGeom: Column, srcCRSCol: Column, dstCRS: CRS): TypedColumn[Any, Geometry] =
    ReprojectGeometry(sourceGeom, srcCRSCol, dstCRS)

  /** Reproject a column of geometry from one CRS to another. */
  def reproject_geometry(sourceGeom: Column, srcCRS: CRS, dstCRS: CRS): TypedColumn[Any, Geometry] =
    ReprojectGeometry(sourceGeom, srcCRS, dstCRS)

  /** Render Tile as ASCII string, for debugging purposes. */
  def render_ascii(col: Column): TypedColumn[Any, String] =
    DebugRender.RenderAscii(col)

  /** Render Tile cell values as numeric values, for debugging purposes. */
  def render_matrix(col: Column): TypedColumn[Any, String] =
    DebugRender.RenderMatrix(col)

  /** Cellwise less than value comparison between two tiles. */
  def local_less(left: Column, right: Column): TypedColumn[Any, Tile] =
    Less(left, right)

  /** Cellwise less than value comparison between a tile and a scalar. */
  def local_less[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    Less(tileCol, value)

  /** Cellwise less than or equal to value comparison between a tile and a scalar. */
  def local_less_equal(left: Column, right: Column): TypedColumn[Any, Tile]  =
    LessEqual(left, right)

  /** Cellwise less than or equal to value comparison between a tile and a scalar. */
  def local_less_equal[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    LessEqual(tileCol, value)

  /** Cellwise greater than value comparison between two tiles. */
  def local_greater(left: Column, right: Column): TypedColumn[Any, Tile] =
    Greater(left, right)

  /** Cellwise greater than value comparison between a tile and a scalar. */
  def local_greater[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    Greater(tileCol, value)

  /** Cellwise greater than or equal to value comparison between two tiles. */
  def local_greater_equal(left: Column, right: Column): TypedColumn[Any, Tile]  =
    GreaterEqual(left, right)

  /** Cellwise greater than or equal to value comparison between a tile and a scalar. */
  def local_greater_equal[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    GreaterEqual(tileCol, value)

  /** Cellwise equal to value comparison between two tiles. */
  def local_equal(left: Column, right: Column): TypedColumn[Any, Tile]  =
    Equal(left, right)

  /** Cellwise equal to value comparison between a tile and a scalar. */
  def local_equal[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    Equal(tileCol, value)

  /** Cellwise inequality comparison between two tiles. */
  def local_unequal(left: Column, right: Column): TypedColumn[Any, Tile]  =
    Unequal(left, right)

  /** Cellwise inequality comparison between a tile and a scalar. */
  def local_unequal[T: Numeric](tileCol: Column, value: T): TypedColumn[Any, Tile] =
    Unequal(tileCol, value)

  /** Round cell values to nearest integer without chaning cell type. */
  def round(tileCol: Column): TypedColumn[Any, Tile] =
    Round(tileCol)

  /** Take natural logarithm of cell values. */
  def log(tileCol: Column): TypedColumn[Any, Tile] =
    Log(tileCol)

  /** Take base 10 logarithm of cell values. */
  def log10(tileCol: Column): TypedColumn[Any, Tile] =
    Log10(tileCol)

  /** Take base 2 logarithm of cell values. */
  def log2(tileCol: Column): TypedColumn[Any, Tile] =
    Log2(tileCol)

  /** Natural logarithm of one plus cell values. */
  def log1p(tileCol: Column): TypedColumn[Any, Tile] =
      Log1p(tileCol)

  /** Exponential of cell values */
  def exp(tileCol: Column): TypedColumn[Any, Tile] =
    Exp(tileCol)

  /** Ten to the power of cell values */
  def exp10(tileCol: Column): TypedColumn[Any, Tile] =
    Exp10(tileCol)

  /** Two to the power of cell values */
  def exp2(tileCol: Column): TypedColumn[Any, Tile] =
    Exp2(tileCol)

  /** Exponential of cell values, less one*/
  def expm1(tileCol: Column): TypedColumn[Any, Tile] =
    ExpM1(tileCol)

  /** Resample tile using nearest-neighbor */
  def resample[T: Numeric](tileCol: Column, value: T) = Resample(tileCol, value)

  /** Resample tile using nearest-neighbor */
  def resample(tileCol: Column, column2: Column) = Resample(tileCol, column2)

}
